### PR TITLE
Implement Tgz#name

### PR DESCRIPTION
**Pull request by Sammers21**
I have tried to use:

- [eo-yam](https://github.com/decorators-squad/eo-yaml)l, before, but it seems that it is [broken](https://github.com/decorators-squad/eo-yaml/issues/340)
- [jackson-yaml](https://github.com/FasterXML/jackson-dataformats-text/tree/master/yaml) - the API too low level

and, finally, decided to use [snakeyaml](https://bitbucket.org/asomov/snakeyaml/wiki/Documentation) just because it is the best Yaml library for java 

_Closes #20

### CODE REVIEW

**Comment by Sammers21** [comment]
@g4s8, pay attention to the pull request

**Comment by 0crat** [comment]
This pull request [#23](https://github.com/artipie/helm-adapter/issues/23) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/3632fb33-18af-4464-903f-890217704531); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: pom.xml [hunk 1]
```diff
@@ -108,5 +120,26 @@ SOFTWARE.
...
+                <exclude>duplicatefinder:.*</exclude>
+                <exclude>xml:/src/it/settings.xml</exclude>
+                <exclude>dependencies:.*</exclude>
+                <exclude>checkstyle:.*/src/test/resources/.*</exclude>
+              </excludes>
+            </configuration>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
```
Comment (victornoel) [inline pom.xml:143]: @Sammers21 please remove this as it is out of scope of this PR and you can either create an issue so that qulice is configured (remember that there is a parent pom in the artipie project that takes care of setting it up) or add a todo here.

File: src/main/java/com/artipie/helm/TgzArchive.java [hunk 2]
```diff
@@ -61,10 +68,22 @@
...
-        throw new IllegalStateException("Not Implemented");
+        try {
+            final TarArchiveInputStream taris = new TarArchiveInputStream(
+                new GzipCompressorInputStream(new ByteArrayInputStream(this.content))
+            );
+            TarArchiveEntry entry;
+            while ((entry = taris.getNextTarEntry()) != null) {
+                if (!entry.getName().endsWith("Chart.yaml")) {
+                    continue;
+                }
```
Comment (victornoel) [inline src/main/java/com/artipie/helm/TgzArchive.java:79]: @Sammers21 please implement an object whose behaviour is to find the file in an archive (or return optional.empty for example) so that the current method can rely on it only do its job of reading the file or failing

File: src/main/java/com/artipie/helm/TgzArchive.java [hunk 3]
```diff
@@ -61,10 +68,22 @@
      * @return How the archive should be named on the file system
      */
     public String name() {
-        // @todo #12:30min TgzArchive name method
-        //  For now this method is not implemented. We should decompress the archive, find the chart
-        //  version and name, and, return it, binded together.
-        throw new IllegalStateException("Not Implemented");
+        try {
```
Comment (victornoel) [inline src/main/java/com/artipie/helm/TgzArchive.java:71]: @Sammers21 please add at least one test for this method and a todo for writing some more

Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java:71]: Did it in a separated pull request

Comment (victornoel) [reply src/main/java/com/artipie/helm/TgzArchive.java:71]: @Sammers21 ok, and can you update your comment so that it says who this is adressed too as per the policy (see http://www.zerocracy.com/policy.html#42), it has an effect on my quality assessment :)

File: src/main/java/com/artipie/helm/TgzArchive.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java:79]: @victornoel, this would overcomplicate things.

**Comment by victornoel** [comment]
@g4s8 let's refuse this PR, it's too complex and DEV doesn't want to change based on REV requests

File: src/main/java/com/artipie/helm/TgzArchive.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java:80]: @victornoel, a few reasons why this would overcomplicate:
-  the algorithm tries to find a file by it's ending, not by full name
-  the algorithm would not reuse it future
-  the algorithm is just a few lines of code and can be easily understood.

**Comment by Sammers21** [comment]
@victornoel, No, I did not say that "I don't want", I said "I would not".

File: src/main/java/com/artipie/helm/TgzArchive.java (hunk 2 above)
Comment (victornoel) [reply src/main/java/com/artipie/helm/TgzArchive.java:80]: @Sammers21 reuse is not the only reason why you would make an object, the object is to isole responsibilities. It is easier to write tests for it, etc. I won't be explaining what proper OO is here, I believe you are aware of this whole elegant object movement? I believe that this would be good for the project to isolate that functionality in an object. If you don't want to do it, you can always write a todo, if @g4s8 accept it.

Comment (victornoel) [reply src/main/java/com/artipie/helm/TgzArchive.java:80]: @Sammers21 and by the way, I'm pretty sure that `Chart.yaml` won't be the only thing that will be read from this archive. For example in helm there are requirments that can be important, etc.

Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java:80]: @victornoel, `Chart.yaml` is the only file we need for `Index.yaml` generation

Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java:80]: @victornoel, The whole object is already the isolation of the functionality.

**Comment by victornoel** [comment]
@g4s8 I will let you decide what to do based on the above discussions

File: src/main/java/com/artipie/helm/TgzArchive.java (hunk 2 above)
Comment (g4s8) [reply src/main/java/com/artipie/helm/TgzArchive.java:80]: @Sammers21 I agree with @victornoel - this object is doing a lot of different things in a one place:
 - it's parsing the data from tgz format
 - it's searching for some magic file inside it
 - it's parsing yaml to find a name 

It can be simplified in a few different ways, e.g. you can return file content by name: `new TgzArchive(data).entry("Chart.yaml")` and then parse it to find the version. Or move the parsing logic into separate object as @victornoel suggested, e.g.: `String version = new TgzArchive(data).read("Chart.yml", new ChartVersion())`, where chart version will be responsible to parse the version from binary content

Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java:80]: @g4s8, review the PR once again

File: src/main/java/com/artipie/helm/ChartYaml.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+package com.artipie.helm;
+
+import java.util.Map;
+
+/**
+ * The Chart.yaml file.
+ *
+ * @since 0.2
+ */
+public class ChartYaml {
```
Comment (victornoel) [inline src/main/java/com/artipie/helm/ChartYaml.java:33]: @Sammers21 this object is useless, it's only some nice names for getting data from a map, I think you misunderstood the two proposed solution from @g4s8. The idea is to separate getting a file from the tgz into a self-sufficient object.

Comment (Sammers21) [reply src/main/java/com/artipie/helm/ChartYaml.java:33]: @victornoel , TgzArchive is already about it. Why would we need one more object?

Comment (victornoel) [reply src/main/java/com/artipie/helm/ChartYaml.java:33]: @Sammers21 `TgzArchive` has methods pertaining to helm and retrieving the name of the chart. What we propose here is to have something similar to the following:
- `TgzArchive(data).entry("Chart.yaml")` to retrieve the file
- `TgzHelmChart(data).name()` that returns the name of the chart using `TgzArchive` to find the file and snakeyaml to parse it, no need to introduce a class just to wrap a `Map`.
- also like this, we can have an interface `HelmChart` with the method `name()` and `TgzHelmChart` implements it.

Comment (Sammers21) [reply src/main/java/com/artipie/helm/ChartYaml.java:33]: It does not make sense to be because it is the same as `TgzHelmChart` as `TgzArchive`

Comment (Sammers21) [reply src/main/java/com/artipie/helm/ChartYaml.java:33]: > also like this, we can have an interface HelmChart with the method name() and TgzHelmChart implements it.

the role of this pull request is just to implement a single method, NOT to reinvent the whole design. think about it

Comment (victornoel) [reply src/main/java/com/artipie/helm/ChartYaml.java:33]: @Sammers21 you are right about your last point, it's out of scope, I'm not proposing to do it, just to give some context. I stand by the rest of my comment, see with @g4s8 if he accepts the current design, but I believe the current state of the PR is not acceptable.

Comment (g4s8) [reply src/main/java/com/artipie/helm/ChartYaml.java:33]: @Sammers21 I agree with @victornoel here - this class doesn't make any sense, because it's just getting the map value by key. We're proposing a bit different thing: separate the logic of tgz parsing and the logic of yaml parsing, you can implement it as you want, but it's too completely different things. It may be done with many ways, e.g. `new ChartYml(new TgzArchive(bytes).item("chart.yml")).version()` - `ChartYml` is responsible to parse `Chart.yml` file, `TgzArchive` is responsible for parsing tgz archives.

Comment (Sammers21) [reply src/main/java/com/artipie/helm/ChartYaml.java:35]: @g4s8 , do you like it better now?

**Comment by 0crat** [comment]
Job was finished in 20 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by 0crat** [comment]
> @0crat quality good ([here](https://github.com/artipie/helm-adapter/issues/23#issuecomment-636094653))

@sereshqua There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 4@be20caf8-36de-456b-b126-0cbe9e43e13b, thread: PQ-CT2E6TK9B
com.zerocracy.farm.reactive.StkRuntime[76] java.lang.IllegalStateException: org.codehaus.groovy.runtime.InvokerInvocationException in com.zerocracy.stk.pm.cost.pay_cash
org.codehaus.groovy.reflection.CachedMethod[105] org.codehaus.groovy.runtime.InvokerInvocationException: java.io.IOException: Failed to add transaction
com.zerocracy.pm.cost.Ledger[133] java.io.IOException: Failed to add transaction
org.apache.commons.dbcp.BasicDataSource[1549] org.apache.commons.dbcp.SQLNestedException: Cannot create PoolableConnectionFactory (SSL error: Remote host closed connection during handshake)
org.postgresql.ssl.MakeSSL[67] org.postgresql.util.PSQLException: SSL error: Remote host closed connection during handshake
sun.security.ssl.SSLSocketImpl[1009] javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake
sun.security.ssl.InputRecord[505] java.io.EOFException: SSL peer shut down incorrectly
```

[0.53.15](https://github.com/zerocracy/farm): CID: [14c3f7d9-95a2-4d8d-8a55-6bb740992e64](https://www.0crat.com/footprint/CT2E6TK9B/14c3f7d9-95a2-4d8d-8a55-6bb740992e64), Type: "Pay cash"
